### DISCUSSION TITLE
[NOTICE] Increase your ARP table size to scale beyond ~40 containers

### DISCUSSION THREAD

**Discussion by raulk** [testground/testground#1240] [General]
If you are having network troubles when running Testground container workloads of over ~40 containers, this post is relevant to you. Visible effects are:

- IO timeouts when dialling peers.
- High latency jitter.
- General network flakiness.

We have identified that Linux distros ship with too low ARP cache sizes by default, and the Docker bridge network driver seems to rely heavily on the ARP cache.

## How to identify if you're affected

1. Check your `dmesg` logs while running a Testground test with more than ~40 containers that connect to one another. You can also use the `storm` test case from the `benchmarks` test plan [in the Testground repo](https://github.com/testground/testground/blob/master/plans/benchmarks/compositions/storm.toml).
    
    If you observe messages like this, you are affected.

    ```log
    $ dmesg
    [...]
    [5952197.195796] net_ratelimit: 41 callbacks suppressed
    [5952197.195797] neighbour: arp_cache: neighbor table overflow!
    [5952197.195829] neighbour: arp_cache: neighbor table overflow!
    [5952197.196008] neighbour: arp_cache: neighbor table overflow!
    [5952197.197958] neighbour: arp_cache: neighbor table overflow!
    [5952197.201034] neighbour: arp_cache: neighbor table overflow!
    [5952197.201332] neighbour: arp_cache: neighbor table overflow
    ```

    These are indicating overflow/saturation of the ARP cache. Confirm the sizing of the cache:

    ```sh
    $ sysctl -a | grep gc_thresh | grep neigh
    net.ipv4.neigh.default.gc_thresh1 = 128
    net.ipv4.neigh.default.gc_thresh2 = 512
    net.ipv4.neigh.default.gc_thresh3 = 1024
    net.ipv6.neigh.default.gc_thresh1 = 128
    net.ipv6.neigh.default.gc_thresh2 = 512
    net.ipv6.neigh.default.gc_thresh3 = 1024
    ```

**These are low values.**

## How to fix the issue

_From https://www.packetfence.org/support/faq/solving-neighbour-table-overflow-errors-large-subnets.html_

1. Edit `/etc/sysctl.conf` (or use the `/etc/sysctl.d` facility) to increase the values of ipv4 (and and potentially ipv6, if applicable):

    ```sh
    net.ipv4.neigh.default.gc_thresh1 = 2048
    net.ipv4.neigh.default.gc_thresh2 = 4096 
    net.ipv4.neigh.default.gc_thresh3 = 8192
    ```

2. Apply the changes:

    ```sh
    $ sudo sysctl -p
    ```

**Comment by tchardin** [comment]
Bumped the values as below in order to scale to 200 instances. This works great!
```
net.ipv4.neigh.default.gc_thresh1 = 4096
net.ipv4.neigh.default.gc_thresh2 = 8192
net.ipv4.neigh.default.gc_thresh3 = 16384
```
